[PATCH v2] api: dma: support source free and destination allocation features

VamsiKrishnaA99 opened the pull request:
Following patches outline the required fields in DMA spec to support source segment free and destination segment allocation features. These features are mainly supported for data format type odp_packet_t.

TuomasTaipale commented:
With `--enable-abi-compat`, there seems to be failures. Seems to be fixable by swapping to include `pool_types.h` at line 135.
```diff
@@ -132,7 +132,7 @@ typedef struct odp_dma_pool_param_t {
 } odp_dma_pool_param_t;
 
 /* Includes pool_types.h, which depends on odp_dma_pool_param_t. */
-#include <odp/api/queue_types.h>
+#include <odp/api/pool_types.h>
 
 /**
  * DMA transfer direction
```

TuomasTaipale reviewed (commented):
Thanks for the patches.

A few typos in commit messages:
- "outines" -> "outlines"
- "segements" -> "segments"

VamsiKrishnaA99 commented on include/odp/api/spec/dma_types.h:
```diff
@@ -397,6 +411,14 @@ typedef struct odp_dma_seg_t {
 
+		/** Segment index
+		 *
+		 *  Defines a packet handle index with in the odp_dma_transfer_param_t::dst_seg
+		 *  table to which this segment belongs to. When destination segment allocation
+		 *  support is used, this index field can be used to merge multiple segments
+		 *  into a single packet handle.
```
Thanks Tuomas. From my understanding, above len field denotes the total len of a packet in the to-be-allocated packet array right. Implementation could compute this length field right (add up all the odp_dma_seg_t::len fields of the same indexes). Can you please elaborate if you see any other use case for the new length field.

For example: 
new structure::
struct {
      uint16_t index;
      uint32_t len;
} pkt_handle;  

odp_dma_transfer_param_t::num_dst = 5. 

dst_seg[0].len = 256, dst_seg[1].len = 512, dst_seg[2].len = 64, dst_seg[3] = 32, dst_seg[4] = 256.

dst_seg[0].pkt_handle.index = 0, dst_seg[1].pkt_handle.index = 0, dst_seg[2].len.pkt_handle.index = 1, dst_seg[3].pkt_handle.index = 1, dst_seg[4].pkt_handle.len = 2.

dst_seg[0].pkt_handle.len = 768, dst_seg[1].pkt_handle.len = 0, dst_seg[2].len.pkt_handle.len = 96, dst_seg[3].pkt_handle.len = 0, dst_seg[4].pkt_handle.len = 256.

Result struct will be
odp_dma_result_t::dst_pkt[0] len:: 768, odp_dma_result_t::dst_pkt[1] len:: 96, odp_dma_result_t::dst_pkt[2] len:: 256

VamsiKrishnaA99 replied:
Yes, Thanks Tuomas, it makes sense to have it for pipelining purposes. Will submit the next version with all the corrections.

VamsiKrishnaA99 commented:
Thanks Tuomas for the review comments. I addressed all the comments, will push the next version once the proposal for the new index struct is agreed upon.

TuomasTaipale commented:
Closing this, continues in #2179
